File: Compiler/Syntax/Type.py

class GenericVarDeclaration(object):
    def __init__(self, name):
        self.name = name
        self.deps = []

class GenericConstraint(object):
    def __init__(self):
        self.constraint = None

class GenericDeclaration(object):
    def __init__(self):
        self.generics = []
        self.constraints = []

class Named(object):
    def __init__(self, name, args):
        self.name = name
        self.args = args

    def __str__(self) -> str:
        if len(self.args) == 0:
            return "%s" % self.name
        else:
            args = []
            for a in self.args:
                args.append(str(a))
            return "%s[%s]" % (self.name, ", ".join(args))

    def __repr__(self) -> str:
        return self.__str__()

    def __eq__(self, other) -> bool:
        if isinstance(other, Named):
            return self.name == other.name and self.args == other.args
        else:
            return False

    def __hash__(self) -> int:
        return self.name.__hash__()

class Function(object):
    def __init__(self, args, result):
        self.params = args
        self.result = result

    def __str__(self) -> str:
        args = []
        for a in self.params:
            args.append(str(a))
        return "fn(%s) -> %s" % (", ".join(args), self.result)

    def __repr__(self) -> str:
        return self.__str__()

    def __eq__(self, other) -> bool:
        if isinstance(other, Function):
            return self.result == other.result and self.params == other.params
        else:
            return False

    def __hash__(self) -> int:
        return self.result.__hash__()

class Tuple(object):
    def __init__(self, items):
        self.items = items

    def __str__(self):
        items = []
        for i in self.items:
            items.append(str(i))
        return "(%s)" % ", ".join(items)

    def __repr__(self) -> str:
        return self.__str__()

    def __eq__(self, other) -> bool:
        if isinstance(other, Tuple):
            return self.items == other.items
        else:
            return False

    def __hash__(self) -> int:
        return self.items.__hash__()

class Var(object):
    def __init__(self, name):
        self.name = name

    def __str__(self):
        return "%s" % self.name

    def __repr__(self) -> str:
        return self.__str__()

    def __eq__(self, other) -> bool:
        if isinstance(other, Var):
            return self.name == other.name
        else:
            return False

    def __hash__(self) -> int:
        return self.name.__hash__()

class Type(object):
    def __init__(self, kind):
        self.kind = kind

    def __str__(self) -> str:
        return self.kind.__str__()

    def __repr__(self) -> str:
        return self.__str__()

    def __eq__(self, other) -> bool:
        return self.kind == other.kind
    
    def __hash__(self) -> int:
        return self.kind.__hash__()

File: compiler/src/siko/syntax/Type.rs
use std::fmt;

use super::Identifier::Identifier;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Named(Identifier, Vec<Type>),
    Tuple(Vec<Type>),
    Function(Vec<Type>, Box<Type>),
    Reference(Box<Type>),
    Ptr(Box<Type>),
    SelfType,
    Never,
    NumericConstant(String),
    Void,
    VoidPtr,
}

impl fmt::Display for Type {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Type::Named(id, args) => {
                if args.is_empty() {
                    write!(f, "{}", id)
                } else {
                    let args = args.iter().map(|arg| format!("{}", arg)).collect::<Vec<_>>().join(", ");
                    write!(f, "{}[{}]", id, args)
                }
            }
            Type::Tuple(elements) => {
                let elements = elements
                    .iter()
                    .map(|el| format!("{}", el))
                    .collect::<Vec<_>>()
                    .join(", ");
                write!(f, "({})", elements)
            }
            Type::Function(params, ret) => {
                let params = params
                    .iter()
                    .map(|param| format!("{}", param))
                    .collect::<Vec<_>>()
                    .join(", ");
                write!(f, "fn({}) -> {}", params, ret)
            }
            Type::Reference(inner) => write!(f, "&{}", inner),
            Type::Ptr(inner) => write!(f, "*{}", inner),
            Type::SelfType => write!(f, "Self"),
            Type::Never => write!(f, "!"),
            Type::NumericConstant(value) => write!(f, "{}", value),
            Type::Void => write!(f, "void"),
            Type::VoidPtr => write!(f, "void*"),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TypeParameterDeclaration {
    pub params: Vec<Identifier>,
    pub constraints: Vec<Constraint>,
}

impl fmt::Display for TypeParameterDeclaration {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let params = self
            .params
            .iter()
            .map(|param| format!("{}", param))
            .collect::<Vec<_>>()
            .join(", ");

        if self.constraints.is_empty() {
            write!(f, "[{}]", params)
        } else {
            let constraints = self
                .constraints
                .iter()
                .map(|constraint| format!("{}", constraint))
                .collect::<Vec<_>>()
                .join(", ");
            write!(f, "[{}: {}]", params, constraints)
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ConstraintArgument {
    Type(Type),
    AssociatedType(Identifier, Type),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Constraint {
    pub name: Identifier,
    pub args: Vec<ConstraintArgument>,
}

impl fmt::Display for ConstraintArgument {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ConstraintArgument::Type(ty) => write!(f, "{}", ty),
            ConstraintArgument::AssociatedType(id, ty) => {
                write!(f, "{} = {}", id, ty)
            }
        }
    }
}

impl fmt::Display for Constraint {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let args = self
            .args
            .iter()
            .map(|arg| format!("{}", arg))
            .collect::<Vec<_>>()
            .join(", ");
        write!(f, "{}[{}]", self.name, args)
    }
}
